zooniverse/front-end-monorepo
# Fix QuickTalk Authentication

## PR Overview

Package: fix is _in_ `lib-classifier`, but fix is _for_ `app-project`

This PR fixes an issue where logged-in users couldn't post comments via QuickTalk on _app-project_

Issue:
- when the QuickTalk feature is tested on `lib-classifier`, logged-in users can post comments to the Talk (Subject discussion) thread quite fine.
- however, when the QuickTalk feature is tested on `app-project`, the "create a Talk post" action fails, with the error message saying the user isn't authorised to do so (error 401)

Analysis:
- Problem seems to stem from this line in [talk-client.js](https://github.com/zooniverse/panoptes-javascript-client/blob/7a627dff983b0e4fdf9dd0979330f239a4442e78/lib/talk-client.js#L11): `talkClient.headers = apiClient.headers`
- For some reason, in lib-classifier, `apiClient.headers.Authorization` exists if the user is logged in. In app-project, `apiClient.headers.Authorization`. Seems like there's a difference in the apiClient between lib-classifier and app-project.

Solution:
- In lack of a more elegant solution, this PR fixes the issue by simply force-updating `talkClient.header` with the correct header values, using the known bearerToken we've been tracking to create the Authorization field.
- [ ] 🔨 Room for Improvement 1: use find a way to use the `fixTalkClientAuth` logic by modifying `talkClient.beforeEachRequest` without making the whole damn thing crash. We can't keep jimmying `fixTalkClientAuth()` manually before each request.
- [ ] 🔨 Room for Improvement 2: this may require that the QuickTalkContainer itself be rewritten as a functional component to keep it in line with current FEM standards.

Notes:
- QuickTalk is a feature on the Classifier that only appears with the QuickTalk experimental_tool enabled.
- User has to be logged in to submit a comment, but can be logged out to view comments.
- Some UI improvements are incoming in a separate PR: 1. QuickTalk button (on lower right of screen) needs more visibility, 2. a message saying "You must be logged in to post a comment" needs to be added for better UX.

### How to Review

- Be sure to run `yarn bootstrap` before running `app-project`
- You require a project that 1. has QuickTalk enabled as an experimental tool, and 2. has the Talk boards setup.
- Example URL: https://local.zooniverse.org:3000/projects/darkeshard/test-project-2022/classify/workflow/3581
- Login using a Zooniverse account.
- Click on the QuickTalk icon to open the QuickTalk menu.
- Post a comment.
- Expected: your comment should be posted to the Subject Discussion thread successfully. Your comment should be both immediately visible on the QuickTalk menu and visible on the Subject Discussion Talk page should you visit it.

### Status

Not sure if I should mark this ready for review or a work in progress (see 🔨 Room For Improvement) but comment are still welcome.

❗ UPDATE / WARNING: I just realised that on `frontend.preview.zooniverse.org`, **the talkClient will always ping production** (i.e. `talk.zooniverse.org`), **regardless of the ?env=staging** query. This make me think that the disconnect between the `apiClient` in lib-classifier and the `apiClient` in app-project is bigger than previously imagined. Hmm...
